### PR TITLE
Run Smokey tests against the standby CDN (and mirrors) via a proxy

### PR DESCRIPTION
**Pull request by ChrisBAshton**
## What

This PR contains the necessary changes to allow us to run our Smokey suite in different contexts, by virtue of a lightweight proxy that sets the host and other headers as appropriate.

1. The default Smokey test suite will continue to run in exactly the same way as it always has done.
2. We can run Smokey against the failover CDN via the proxy in a `failoverCDN` context. This will run Smokey only against scenarios that don't have the tag `@notcloudfront`.
3. We can run Smokey against any one of our mirrors via the proxy in the relevant context: `mirrorS3`, `mirrorS3Replica` or `mirrorGCS`. This will run Smokey only against scenarios that have the tag `@worksonmirror`.

Try it out (note that `FAILOVER_CDN_HOST` should be replaced by the actual distribution ID that can be found at govuk-dns-tf#69):

```
shell1$ env FAILOVER_CDN_HOST="secret.cloudfront.net" GOVUK_PROXY_PROFILE=failoverCDN ruby govuk_proxy_runner.rb

shell2$ env ENVIRONMENT=production GOVUK_PROXY_PROFILE=failoverCDN SIGNON_EMAIL="replaceme" SIGNON_PASSWORD="replaceme" bundle exec cucumber --tags="not @notreplatforming and not @notcloudfront"
```

This PR has been manually tested. See https://argo.eks.integration.govuk.digital/applications/cluster-services/smokey?operation=false&orphaned=false&node=batch%2FJob%2Fapps%2Fsmokey-28432150&tab=logs&resource=

## Why

We want to have continued confidence that our failover CDN is working. As a bonus, the same changes required to enable testing of the failover CDN also allow us to test the mirrors, so let's do that also.

We've had to build our own proxy because [Selenium doesn't support setting request headers](https://stackoverflow.com/a/15647143), and thus requires use of a proxy like browsermob-proxy, to inject headers. I _really_ don't want to use browsermob-proxy - it was already in Smokey a couple of years ago, but was [removed because of its flakiness and Java dependency](https://github.com/alphagov/smokey/pull/924)

Note also that there are some [HTTP request support methods in Smokey](https://github.com/alphagov/smokey/blob/2319cca9a815ff40ba5eed8483bfc4089a04a480/features/support/http_requests.rb), which bypass Capybara and have support for custom headers. I did [investigate reusing those methods](https://github.com/alphagov/smokey/pull/1220) but it would have meant rewriting our entire test suite (to stop
using Capybara).

## Scope

Next steps:

- Configure these new test suites to run somewhere. I'm thinking the failover CDN and all 3 sets of mirror tests should run on a cron, once per day, and should only be concerned with 'production' for now at least.

Not in scope for this PR:
- Fixing broken CDN failover assets (e.g. 404/503'ing assets)
- Consolidating how we run Smokey tests, so that the 'default' tests go through the proxy too. I've started a [spike](https://github.com/alphagov/smokey/pull/1231) into that, but it could take a bit of work.
- Refactoring away the custom HTTP request support methods in `http_requests.rb`

Trello: https://trello.com/c/2XwAz6JD/3365-run-automatic-smokey-tests-against-the-standby-cdn-5

### CODE REVIEW

File: features/support/visiting_pages.rb [hunk 1]
```diff
@@ -7,3 +7,7 @@ def visit_path(path)
     visit "#{path}?smokey_cachebust=#{rand.to_s}"
   end
 end
+
+def wait_for_any_redirects_to_govuk
+  page.has_css?(".govuk-template")
```
Comment (robinjam) [inline features/support/visiting_pages.rb:12]: Could you add a quick comment here summarising what you've said in the commit description?

**Review by robinjam — COMMENTED** [review]
Nice one! Bit annoying that there's no way to override headers in webdriver tests, but the proxy is an elegant workaround.

Just a few small suggestions

File: proxy.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    headers: {
+      "User-Agent": "Smokey Test / Ruby",
+    },
+  }
+}
+
+PROFILE = ENV.has_key?("PROXY_PROFILE") ? PROFILES[ENV["PROXY_PROFILE"].to_sym] : nil
+raise "PROXY_PROFILE ENV var needs to be one of #{PROFILES.keys.join(', ')}" if PROFILE.nil?
+
+class MyProxy < WEBrick::HTTPServlet::AbstractServlet
```
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply proxy.rb:54]: I'm going to call YAGNI on that one 😄 they're not used yet and I'm not sure they ever would be.
The error message for calling an undefined HTTP method should make it quite clear what needs doing, if that scenario does come about.

File: proxy.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  def do_POST(request, response)
+    make_http_request(request, response) do |uri|
+      req = Net::HTTP::Post.new(uri)
+      req.set_form_data(CGI::parse(request.body))
+      req
+    end
+  end
+
+  def make_http_request(request, response)
```
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply proxy.rb:69]: Eurgh, yeah I'm not keen on this method. I'm semi hoping to be able to do away with it all and somehow get the tests that use this method to make a call to the in-house proxy library instead, but that can come later on.

File: proxy.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      http.request(req)
+    end
+
+    response.status = resp.code
+    response.content_type = resp["content-type"]
+    response.body = resp.body
+
+    case resp
+    when Net::HTTPRedirection
+      response["Location"] = URI.parse(resp["location"])
```
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply proxy.rb]: No reason, in theory - just the overhead of having to enumerate all the headers in `resp` and set those headers on `response`. A small risk of some side-effect too, I suppose. The only ones that have made a material difference to the running of the tests so far have been the content type and Location headers.

File: proxy.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply proxy.rb:69]: I'll see if I can give it a refactor 👍

[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply proxy.rb]: 🤮 it's a bit spaghetti, agreed. I've decided to leave it - now that the file is reduced to just the GovukProxy logic (no profiles or 'runner' logic - see latest commit) then it's less daunting to have to read and understand.

**Review by robinjam — APPROVED** [review]
LGTM! 👍

**Pull request merged**